### PR TITLE
Instance metrics can be queried by instance index

### PR DESCRIPTION
**Pull request by qibobo**
[#159756219]

### CODE REVIEW

File: api/lib/routes/metricHelper.js [hunk 1]
```diff
@@ -3,12 +3,22 @@
...
   if (!metricType) {
     return { valid: false, message: "metric_type is required" };
   }
+  if (routeHelper.isMissing(instanceIndex)) {
+    instanceIndex = 0;
+  } else {
+    instanceIndex = parseInt(instanceIndex);
+    if (!routeHelper.isInteger(instanceIndex)) {
+      return { valid: false, message: "instance-index must be an integer" };
+    }
```
Comment (cdlliuy) [inline api/lib/routes/metricHelper.js]: how about instance-index= -1?

File: src/autoscaler/metricscollector/server/metric_handler_test.go [hunk 2]
```diff
@@ -50,6 +50,70 @@ var _ = Describe("MetricHandler", func() {
...
+
+					Expect(err).ToNot(HaveOccurred())
+					Expect(errJson).To(Equal(&models.ErrorResponse{
+						Code:    "Bad-Request",
+						Message: "Incorrect instanceIndex parameter in query string",
+					}))
+				})
+			})
+
+			Context("when instanceindex time is not a number", func() {
```
Comment (cdlliuy) [inline src/autoscaler/metricscollector/server/metric_handler_test.go]: "instanceindex time" ?

File: src/autoscaler/metricscollector/server/metric_handler_test.go [hunk 3]
```diff
@@ -50,6 +50,70 @@ var _ = Describe("MetricHandler", func() {
...
+
+					Expect(err).ToNot(HaveOccurred())
+					Expect(errJson).To(Equal(&models.ErrorResponse{
+						Code:    "Bad-Request",
+						Message: "Error parsing instanceIndex",
+					}))
+				})
+			})
+
+			Context("when instanceindex time is smaller than 0", func() {
```
Comment (cdlliuy) [inline src/autoscaler/metricscollector/server/metric_handler_test.go]: same as above

**Comment by cdlliuy** [comment]
LGTM

File: api/lib/utils/metricsCollectorUtils.js [hunk 4]
```diff
@@ -28,25 +28,26 @@ module.exports = function(metricsCollectorSettings) {
   metricsCollectorUtilObj.getMetricHistory = function getMetricHistory(parameters, callback) {
     var appId = parameters.appId;
     var metricType = parameters.metricType;
+    var instanceIndex = parameters.instanceIndex;
     var startTime = parameters.startTime;
     var endTime = parameters.endTime;
     var order = parameters.order;
-    logger.info("Get metric histories", { "appId": appId, "metricType": metricType, "startTime": startTime, "endTime": endTime, "order": order });
-    var options = getOptions(appId, metricType, startTime, endTime, order);
+    logger.info("Get metric histories", { "appId": appId, "metricType": metricType, "instanceIndex": instanceIndex, "startTime": startTime, "endTime": endTime, "order": order });
```
Comment (boyang9527) [inline api/lib/utils/metricsCollectorUtils.js]: should be debug level

File: api/lib/utils/metricsCollectorUtils.js [hunk 5]
```diff
@@ -28,25 +28,26 @@ module.exports = function(metricsCollectorSettings) {
...
+    var options = getOptions(appId, metricType, instanceIndex, startTime, endTime, order);
     request(options, function(error, response, body) {
       if (error) {
-        logger.error("Error occurred during get metric histories ", { "appId": appId, "metricType": metricType, "startTime": startTime, "endTime": endTime, "order": order, "error": error });
+        logger.error("Error occurred during get metric histories ", { "appId": appId, "metricType": metricType, "instanceIndex": instanceIndex, "startTime": startTime, "endTime": endTime, "order": order, "error": error });
         error.statusCode = HttpStatus.INTERNAL_SERVER_ERROR;
         callback(error, null);
       } else if (response.statusCode === HttpStatus.OK) {
-        logger.info("Get metric histories successfully", { "appId": appId, "metricType": metricType, "startTime": startTime, "endTime": endTime, "order": order });
+        logger.info("Get metric histories successfully", { "appId": appId, "metricType": metricType, "instanceIndex": instanceIndex, "startTime": startTime, "endTime": endTime, "order": order });
```
Comment (boyang9527) [inline api/lib/utils/metricsCollectorUtils.js]: debug level

File: src/autoscaler/db/sqldb/instancemetrics_sqldb.go [hunk 6]
```diff
@@ -115,16 +115,36 @@ func (idb *InstanceMetricsSQLDB) RetrieveInstanceMetrics(appid string, name stri
...
-		return nil, err
+	var rows *sql.Rows
+	if instanceIndex >= 0 {
+		rows1, err := idb.sqldb.Query(queryByInstanceIndex, appid, instanceIndex, name, start, end)
+		if err != nil {
+			idb.logger.Error("failed-retrieve-instancemetrics-from-appinstancemetrics-table", err,
+				lager.Data{"query": query, "appid": appid, "instanceindex": instanceIndex, "metricName": name, "start": start, "end": end, "orderType": orderType})
+			return nil, err
+		}
+		rows = rows1
```
Comment (boyang9527) [inline src/autoscaler/db/sqldb/instancemetrics_sqldb.go]: why use rows1, and rows2 ?

File: src/autoscaler/db/sqldb/instancemetrics_sqldb.go [hunk 7]
```diff
@@ -101,7 +101,7 @@ func (idb *InstanceMetricsSQLDB) SaveMetricsInBulk(metrics []*models.AppInstance
 	return nil
 }
 
-func (idb *InstanceMetricsSQLDB) RetrieveInstanceMetrics(appid string, name string, start int64, end int64, orderType db.OrderType) ([]*models.AppInstanceMetric, error) {
+func (idb *InstanceMetricsSQLDB) RetrieveInstanceMetrics(appid string, instanceIndex int64, name string, start int64, end int64, orderType db.OrderType) ([]*models.AppInstanceMetric, error) {
```
Comment (boyang9527) [inline src/autoscaler/db/sqldb/instancemetrics_sqldb.go]: why use int64 for instance index?

File: src/autoscaler/metricscollector/server/metric_handler.go [hunk 8]
```diff
@@ -35,16 +35,42 @@ func NewMetricHandler(logger lager.Logger, cfc cf.CfClient, consumer noaa.NoaaCo
...
 	orderParam := r.URL.Query()["order"]
 
-	h.logger.Debug("get-metric-histories", lager.Data{"appId": appId, "metrictype": metricType, "start": startParam, "end": endParam, "order": orderParam})
+	h.logger.Debug("get-metric-histories", lager.Data{"appId": appId, "metrictype": metricType, "instanceIndex": instanceIndexParam, "start": startParam, "end": endParam, "order": orderParam})
 
 	var err error
 	start := int64(0)
 	end := int64(-1)
 	order := db.ASC
+	instanceIndex := int64(-1)
```
Comment (boyang9527) [inline src/autoscaler/metricscollector/server/metric_handler.go]: same as above, any special reason to use int64 instead of int?

Comment (qibobo) [reply src/autoscaler/metricscollector/server/metric_handler.go]: @boyang9527 instanceIndex will be the result of strconv.ParseInt which is a int64, so define it int64 here and convert to int when calling db.RetrieveInstanceMetrics.

File: api/lib/utils/metricsCollectorUtils.js (hunk 5 above)
Comment (qibobo) [reply api/lib/utils/metricsCollectorUtils.js]: @boyang9527 there is no debug in node.

**Comment by boyang9527** [comment]
LGTM

**Comment by boyang9527** [comment]
LGTM

**Pull request merged**